File: frontend/src/lib/pdf/pdf-document.ts
import { jsPDF } from "jspdf";

import { FullDocument, ItemTable } from "@/types";
import { formatCurrency } from "@/utils";
import { formatDate } from "@/utils/formatters";

export function generarPDF(doc: Omit<FullDocument, "data"> & { items: ItemTable[] }) {
	const { client, company, document, items } = doc;
	const { tipoDocumento: tipo, numeroDocumento: numero, fecha, iva: tax } = document;

	const pdf = new jsPDF();
	const green = "#006630";
	const dark_green = "#0d542b";
	let y = 10;
	const leftX = 10;
	const rightX = 115;
	const width = 190;
	const ROW_HEIGHT = 10;
	const PAGE_LIMIT = 290;
	const widthTable = 197;

	const addPageIfNeeded = (minSpace: number = 40) => {
		const pageHeight = pdf.internal.pageSize.height;
		if (y + minSpace >= pageHeight - 15) {
			pdf.addPage();
			y = 10;
		}
	};

	pdf.setFillColor(green);
	pdf.setDrawColor(green);
	pdf.rect(10, y, width, 12, "DF");

	pdf.setTextColor(255, 255, 255);
	pdf.setFont("helvetica", "bold");
	pdf.setFontSize(11);
	pdf.addImage("/images/LogoB.png", leftX + 5, y + 2.5, 7, 7);
	pdf.text(`eformas RAP | ${tipo}`, leftX + 13, y + 9);

	pdf.setTextColor(0, 0, 0);

	y += 20;

	pdf.setFont("helvetica", "bold");
	pdf.text(company.name, leftX + 5, y);
	pdf.setFont("helvetica", "normal");

	pdf.addImage("/icons/phone.png", "PNG", leftX + 5, y + 3, 3.5, 3.5);
	pdf.addImage("/icons/mail.png", "PNG", leftX + 5, y + 9.5, 3.5, 3.5);
	pdf.addImage("/icons/address.png", "PNG", leftX + 5, y + 15, 3.5, 3.5);
	pdf.addImage("/icons/web.png", "PNG", leftX + 5, y + 27, 3.5, 3.5);

	pdf.text(company.phone, leftX + 10, y + 6);
	pdf.text(company.email, leftX + 10, y + 12);
	pdf.text(company.address, leftX + 10, y + 18);
	pdf.text(company.city, leftX + 10, y + 24);
	pdf.text(company.web, leftX + 10, y + 30);
	pdf.link(leftX + 5, y + 20, 35, 10, { url: `https://${company.web}` });

	pdf.setFont("helvetica", "bold");
	pdf.text(client.name, rightX, y);
	pdf.setFont("helvetica", "normal");

	pdf.addImage("/icons/phone.png", "PNG", rightX, y + 3, 3.5, 3.5);
	pdf.addImage("/icons/mail.png", "PNG", rightX, y + 9.5, 3.5, 3.5);
	pdf.addImage("/icons/address.png", "PNG", rightX, y + 15, 3.5, 3.5);

	pdf.text(client.phone || "", rightX + 5, y + 6);
	pdf.text(client.email, rightX + 5, y + 12);
	pdf.text(client.address || "", rightX + 5, y + 18);
	pdf.text(client.city || "", rightX + 5, y + 24);
	if (client.nif) {
		pdf.addImage("/icons/id.png", "PNG", rightX, y + 27, 3.5, 3.5);
		pdf.text(client.nif || "", rightX + 5, y + 30);
	}

	y = 70;

	// Info factura

	pdf.setFillColor(green);

	pdf.rect(10, y, width, 10, "DF");
	pdf.rect(10, y + 10, width, 10);

	pdf.setTextColor(255, 255, 255);

	if (tax) {
		pdf.text(`N° ${tipo}`, width / 6, y + 7);
		pdf.text("Fecha", 99, y + 7);
		pdf.text("DNI", 165, y + 7);

		pdf.setDrawColor(dark_green);
		pdf.line(73.5, y, 73.5, y + 10);
		pdf.line(136.5, y, 136.5, y + 10);
		pdf.setDrawColor(green);
		pdf.line(73.5, y + 10, 73.5, y + 20);
		pdf.line(136.5, y + 10, 136.5, y + 20);

		pdf.setTextColor(0, 0, 0);
		pdf.text(numero ?? "", 10 + width / 6 - numero.length, y + 16);
		pdf.text(formatDate(document.fecha), 95, y + 16);
		pdf.text(company.nif, 158, y + 16);
	} else {
		pdf.text(`N° ${tipo}`, width / 4, y + 7);
		pdf.text("Fecha", 147, y + 7);
		pdf.setTextColor(0, 0, 0);
		pdf.setDrawColor(dark_green);
		pdf.line(105, y, 105, y + 10);

		pdf.setDrawColor(green);
		pdf.line(105, y + 10, 105, y + 20);

		pdf.text(numero ?? "", 10 + width / 4 - numero.length, y + 16);
		pdf.text(formatDate(fecha), 142.5, y + 16);
	}

	y = 100;

	// Tabla
	const header = () => {
		pdf.setFillColor(green);
		pdf.rect(10, y, width, 10, "DF");
		pdf.setTextColor(255, 255, 255);
		pdf.setFont("helvetica", "bold");
		pdf.text("Descripción", 15, y + 7);
		pdf.text("Precio", 125, y + 7);
		pdf.text("Ud.", 150, y + 7);
		pdf.text("Importe", 178, y + 7);
		pdf.setDrawColor(dark_green);
		pdf.line(115, y, 115, y + 10);
		pdf.line(145, y, 145, y + 10);
		pdf.line(160, y, 160, y + 10);
		y += 10;
	};

	header();

	pdf.setFont("helvetica", "normal");
	pdf.setTextColor(0, 0, 0);

	items.forEach(({ description, price, quantity, total }) => {
		const lines = pdf.splitTextToSize(description, 95);

		const rowHeight = Math.max(ROW_HEIGHT, lines.length * 5);

		if (y + rowHeight > PAGE_LIMIT) {
			pdf.addPage();
			y = 10;
			header();
			pdf.setFont("helvetica", "normal");
			pdf.setTextColor(0, 0, 0);
		}

		pdf.rect(10, y, width, rowHeight);

		pdf.setDrawColor(green);
		pdf.line(115, y, 115, y + rowHeight);
		pdf.line(145, y, 145, y + rowHeight);
		pdf.line(160, y, 160, y + rowHeight);

		const p = formatCurrency(price);
		const q = quantity.toString();
		const t = formatCurrency(total);

		pdf.text(lines, 15, y + 7);

		if (price) pdf.text(p, 130 - p.length, y + 7);
		if (quantity) pdf.text(q, 152 - q.length, y + 7);
		pdf.text(t, widthTable - t.length * 2, y + 7);

		y += rowHeight;
	});

	const base = items.reduce((sum, { total }) => (sum += total), 0);
	const subtotal = formatCurrency(base);
	const iva = formatCurrency(base * 0.21);
	const total = formatCurrency(tax ? base + base * 0.21 : base);

	addPageIfNeeded(60);

	pdf.setFont("helvetica", "bold");
	pdf.setTextColor(0, 0, 0);

	if (tax) {
		pdf.rect(10, y, width, 10);
		pdf.text("BASE IMPONIBLE", 15, y + 7);
		pdf.text(subtotal, widthTable - subtotal.length * 2, y + 7);
		y += 10;

		addPageIfNeeded(40);
		pdf.rect(10, y, width, 10);
		pdf.text("21% IVA", 15, y + 7);
		pdf.text(iva, widthTable - iva.length * 2, y + 7);
		y += 10;
	}
	addPageIfNeeded(40);
	pdf.rect(10, y, width, 10);
	pdf.text("TOTAL", 15, y + 7);
	pdf.text(total, widthTable - total.length * 2, y + 7);
	y += 10;

	const condiciones = "Condiciones de pago:";

	addPageIfNeeded(40);
	pdf.rect(10, y, width, 10);
	pdf.text(condiciones, 15, y + 7);

	pdf.setFont("helvetica", "normal");

	const texto = "Se abonará el 40% al inicio de la obra y el resto al final de la misma.";
	const lineas = pdf.splitTextToSize(texto, width - 40);

	pdf.text(lineas, 16 + condiciones.length * 2, y + 7);

	return pdf;
}

File: backend/Utils/PdfDocument.cs
using QuestPDF.Fluent;
using QuestPDF.Helpers;
using QuestPDF.Infrastructure;

namespace ReformasRapBackend.Utils;

public class PdfDocument(PdfDoc doc) : IDocument
{
    private const string Green = "#006630";
    private const string DarkGreen = "#0d542b";
    private const float Border = 0.20F;
    private readonly string _baseDir = AppContext.BaseDirectory;

    public void Compose(IDocumentContainer container)
    {
        container.Page(page =>
        {
            page.Margin(40);
            page.Content().Element(Content);
            page.Footer().Element(Footer);
        });
    }

    private void Header(IContainer container)
    {
        container.Column(col =>
        {
            col.Item()
                .Height(1, Unit.Centimetre)
                .Border(Border, Unit.Millimetre)
                .BorderColor(Green)
                .Background(Green)
                .Row(row =>
                {
                    var imagePath = Path.Combine(_baseDir, "Resources", "Images", "LogoB.png");
                    row.ConstantItem(10);
                    row.ConstantItem(7, Unit.Millimetre).AlignMiddle().Image(imagePath);
                    row.RelativeItem()
                        .PaddingBottom(4)
                        .AlignBottom()
                        .Text($"eformas RAP | {doc.InfoDocument.Tipo}")
                        .FontColor(Colors.White).Bold();
                });

            col.Item().Height(10);
            col.Item().Row(InfoHeader);

            col.Item().Height(10);
            col.Item().Row(InfoFactura);
        });
    }

    private void InfoHeader(RowDescriptor row)
    {
        var company = doc.CompanyInfo;
        var client = doc.ClientInfo;

        row.RelativeItem()
            .PaddingLeft(20)
            .PaddingRight(10)
            .Column(c =>
            {
                c.Item().Text(company.Name).Bold();
                c.Item().IconText(ImagePath( "phone"), company.Phone);
                c.Item().IconText(ImagePath("mail"), company.Email);
                c.Item().IconText(ImagePath("address"), company.Address);
                c.Item().PaddingLeft(16).Text(company.City);
                c.Item().IconText(ImagePath("web"), company.Web);
            });

        row.RelativeItem()
            .PaddingRight(20)
            .PaddingLeft(10)
            .Column(c =>
            {
                c.Item().Text(client.Name).Bold();
                c.Item().IconText(ImagePath("phone"), client.Phone);
                c.Item().IconText(ImagePath("mail"), client.Email);
                c.Item().IconText(ImagePath("address"), client.Address);
                c.Item().PaddingLeft(16).Text(client.City);
                if (!string.IsNullOrEmpty(client.Nif))
                {
                    c.Item().IconText(ImagePath("id"), client.Nif);
                }
            });
        return;

        string ImagePath(string icon) => Path.Combine(_baseDir, "Resources", "Images", $"{icon}.png");
    }
    
    private void InfoFactura(RowDescriptor row)
    {
        row.RelativeItem().Column(col =>
        {
            col.Item()
                .Height(8, Unit.Millimetre)
                .Border(Border, Unit.Millimetre)
                .BorderColor(DarkGreen)
                .Background(Green)
                .AlignCenter()
                .AlignMiddle()
                .Text($"Nº {doc.InfoDocument.Tipo}")
                .FontColor(Colors.White);

            col.Item()
                .Height(8, Unit.Millimetre)
                .Border(Border, Unit.Millimetre)
                .BorderColor(DarkGreen)
                .AlignCenter().AlignMiddle()
                .Text(doc.InfoDocument.Numero);
        });

        row.RelativeItem().Column(col =>
        {
            col.Item()
                .Height(8, Unit.Millimetre)
                .Border(Border, Unit.Millimetre)
                .BorderColor(DarkGreen)
                .Background(Green)
                .AlignCenter()
                .AlignMiddle()
                .Text("Fecha")
                .FontColor(Colors.White);

            col.Item()
                .Height(8, Unit.Millimetre)
                .Border(Border, Unit.Millimetre)
                .BorderColor(DarkGreen)
                .AlignCenter().AlignMiddle()
                .Text(doc.InfoDocument.Fecha.ToShortDateString());
        });

        if (doc.InfoDocument.Iva)
        {
            row.RelativeItem().Column(col =>
            {
                col.Item()
                    .Height(8, Unit.Millimetre)
                    .Border(Border, Unit.Millimetre)
                    .BorderColor(DarkGreen)
                    .Background(Green)
                    .AlignCenter()
                    .AlignMiddle()
                    .Text("DNI")
                    .FontColor(Colors.White);

                col.Item()
                    .Height(8, Unit.Millimetre)
                    .Border(Border, Unit.Millimetre)
                    .BorderColor(DarkGreen)
                    .AlignCenter().AlignMiddle()
                    .Text(doc.InfoDocument.Nif);
            });
        }
    }

    private void Content(IContainer container)
    {
        container.Column(col =>
        {
            col.Item().Element(Header);
            col.Item().Height(1, Unit.Centimetre);
            col.Item().Element(CreateTable);
            col.Item().Element(CreateTableTotals);
            col.Item().Element(Condiciones);
        });
    }

    private void CreateTable(IContainer container)
    {
        container
            .Border(Border, Unit.Millimetre)
            .BorderColor(DarkGreen)
            .Table(table =>
            {
                table.ColumnsDefinition(columns =>
                {
                    columns.RelativeColumn(1.5F);
                    columns.RelativeColumn(0.4F);
                    columns.RelativeColumn(0.2F);
                    columns.RelativeColumn(0.4F);
                });

                table.Header(h =>
                {
                    h.Cell().Element(CellStyle).Text("Descripción").FontColor(Colors.White);
                    h.Cell().Element(CellStyle).Text("Precio").FontColor(Colors.White);
                    h.Cell().Element(CellStyle).Text("Ud.").FontColor(Colors.White);
                    h.Cell().Element(CellStyleEnd).Text("Importe").FontColor(Colors.White);
                    return;

                    static IContainer CellStyle(IContainer container) =>
                        container
                            .DefaultTextStyle(t => t.SemiBold())
                            .Height(8, Unit.Millimetre)
                            .Background(Green)
                            .BorderRight(Border, Unit.Millimetre)
                            .BorderColor(DarkGreen)
                            .AlignCenter()
                            .AlignMiddle();

                    static IContainer CellStyleEnd(IContainer container) =>
                        container
                            .DefaultTextStyle(t => t.SemiBold())
                            .Height(8, Unit.Millimetre)
                            .Background(Green)
                            .BorderColor(DarkGreen)
                            .AlignCenter()
                            .AlignMiddle();
                });

                foreach (var item in doc.Items)
                {
                    table.Cell().Element(CellStyle).AlignLeft().Text(item.Description);
                    table.Cell().Element(CellStyle).Text(FormatCurrency(item.Price));
                    table.Cell().Element(CellStyle).AlignCenter().Text(item.Quantity.ToString());
                    table.Cell().Element(CellStyleEnd).Text(FormatCurrency(item.Total));
                    continue;

                    static IContainer CellStyle(IContainer container) =>
                        container
                            .BorderBottom(Border, Unit.Millimetre)
                            .BorderRight(Border, Unit.Millimetre)
                            .BorderColor(DarkGreen)
                            .Padding(5)
                            .AlignRight();

                    static IContainer CellStyleEnd(IContainer container) =>
                        container
                            .BorderBottom(Border, Unit.Millimetre)
                            .BorderColor(DarkGreen)
                            .Padding(5)
                            .AlignRight();
                }
            });
    }

    private void CreateTableTotals(IContainer container)
    {
        container
            .BorderBottom(Border, Unit.Millimetre)
            .BorderColor(DarkGreen).Table(table =>
            {
                table.ColumnsDefinition(columns =>
                {
                    columns.RelativeColumn();
                    columns.RelativeColumn(0.4F);
                });

                table.Cell().Element(FilaTotalsL).Text("BASE IMPONIBLE").Bold();
                table.Cell().Element(FilaTotalsR).AlignRight().Text(FormatCurrency(doc.Totals.Subtotal)).Bold();
                table.Cell().Element(FilaTotalsL).Text("21% IVA").Bold();
                table.Cell().Element(FilaTotalsR).AlignRight().Text(FormatCurrency(doc.Totals.Iva)).Bold();
                table.Cell().Element(FilaTotalsL).Text("TOTAL").Bold();
                table.Cell().Element(FilaTotalsR).AlignRight().Text(FormatCurrency(doc.Totals.Total)).Bold();

                return;

                static IContainer FilaTotalsL(IContainer container) =>
                    container
                        .BorderBottom(Border, Unit.Millimetre)
                        .BorderLeft(Border, Unit.Millimetre)
                        .BorderColor(DarkGreen)
                        .PaddingVertical(5)
                        .PaddingHorizontal(5);

                static IContainer FilaTotalsR(IContainer container) =>
                    container
                        .BorderBottom(Border, Unit.Millimetre)
                        .BorderRight(Border, Unit.Millimetre)
                        .BorderColor(DarkGreen)
                        .PaddingVertical(5)
                        .PaddingHorizontal(5);
            });
    }

    private void Condiciones(IContainer container)
    {
        container
            .Height(1, Unit.Centimetre)
            .BorderLeft(Border, Unit.Millimetre)
            .BorderRight(Border, Unit.Millimetre)
            .BorderBottom(Border, Unit.Millimetre)
            .BorderColor(DarkGreen)
            .PaddingHorizontal(5)
            .AlignMiddle()
            .Text(t =>
            {
                t.Span("Condiciones de pago: ").Bold();
                t.Span("Se abonara el 40% al inicio de la obra y el resto al final de la misma.");
            });
    }

    private void Footer(IContainer container)
    {
        container.AlignCenter().Text(t =>
        {
            t.CurrentPageNumber();
            t.Span("/");
            t.TotalPages();
        });
    }

    private static string FormatCurrency(decimal value) => $"{Math.Round(value, 2)} €";
}